src/server/dfly_bench.cc

```diff
@@ -0,0 +1,378 @@
+ABSL_FLAG(uint32_t, qps, 20, "QPS schedule at which the generator sends requests to the server");
+ABSL_FLAG(uint32_t, n, 1000, "Number of requests to send per connection");
+ABSL_FLAG(string, h, "localhost", "server hostname/ip");
+ABSL_FLAG(uint64_t, key_minimum, 0, "Min value for keys used");
+ABSL_FLAG(uint64_t, key_maximum, 10'000, "Max value for keys used");
+ABSL_FLAG(string, key_prefix, "key:", "keys prefix");
+ABSL_FLAG(string, key_dist, "U", "U for uniform, N for normal, Z for zipfian");
+ABSL_FLAG(double, zipf_alpha, 0.99, "zipfian alpha parameter");
+ABSL_FLAG(uint64_t, key_stddev, 0,
+          "Standard deviation for non-uniform distribution, 0 chooses"
+          " a default value of (max-min)/6");
+ABSL_FLAG(string, ratio, "1:10", "Set:Get ratio");
+ABSL_FLAG(string, command, "", "custom command with __key__ placeholder for keys");
+
+using namespace std;
+using namespace util;
+using absl::GetFlag;
+using facade::RedisParser;
+using facade::RespVec;
+using tcp = ::boost::asio::ip::tcp;
+
+thread_local absl::InsecureBitGen bit_gen;
+
+class KeyGenerator {
+ public:
+  KeyGenerator(uint32_t min, uint32_t max);
+
+  string operator()();
+
+ private:
+  string prefix_;
+  uint64_t min_, max_, range_;
+  double stddev_ = 1.0 / 6;
+  optional<base::ZipfianGenerator> zipf_;
+  enum DistType { UNIFORM, NORMAL, ZIPFIAN } dist_type_;
+};
+
+class CommandGenerator {
+ public:
+  CommandGenerator(KeyGenerator* keygen);
+
+  string operator()();
+
+ private:
+  KeyGenerator* keygen_;
+  uint32_t ratio_set_ = 0, ratio_get_ = 0;
+  string command_;
+  string cmd_;
+  std::vector<size_t> key_indices_;
+};
+
+CommandGenerator::CommandGenerator(KeyGenerator* keygen) : keygen_(keygen) {
+  command_ = GetFlag(FLAGS_command);
+  if (command_.empty()) {
+    pair<string, string> ratio_str = absl::StrSplit(GetFlag(FLAGS_ratio), ':');
+    CHECK(absl::SimpleAtoi(ratio_str.first, &ratio_set_));
+    CHECK(absl::SimpleAtoi(ratio_str.second, &ratio_get_));
+  } else {
+    for (size_t pos = 0; (pos = command_.find("__key__", pos)) != string::npos; pos += 7) {
+      key_indices_.push_back(pos);
+    }
+  }
+}
+
+string CommandGenerator::operator()() {
+  cmd_.clear();
+  string key;
+  if (command_.empty()) {
+    key = (*keygen_)();
+
+    if (absl::Uniform(bit_gen, 0U, ratio_get_ + ratio_set_) < ratio_set_) {
+      // TODO: value size
+      absl::StrAppend(&cmd_, "set ", key, " val\r\n");
+    } else {
+      absl::StrAppend(&cmd_, "get ", key, "\r\n");
+    }
+  } else {
+    size_t last_pos = 0;
+    for (size_t pos : key_indices_) {
+      key = (*keygen_)();
+      absl::StrAppend(&cmd_, command_.substr(last_pos, pos - last_pos), key);
+      last_pos = pos + 7;
+    }
+    absl::StrAppend(&cmd_, command_.substr(last_pos), "\r\n");
+  }
+  return cmd_;
+}
+
+// Per connection driver.
+class Driver {
+ public:
+  explicit Driver(ProactorBase* p) {
+    socket_.reset(p->CreateSocket());
+  }
+
+  Driver(const Driver&) = delete;
```

Comment:
Nit: you could make this class movable (default should work: `Driver(Driver&&) = default`), which will allow you below to hold a `vector<Driver>` instead of using `unique_ptr`s

Reply:
do not know if it is better but done :)